src/emmy/expression/cse.cljc

```diff
@@ -7,152 +7,154 @@
-;;   replace. At this point, discard all extra bindings (see
-;;   `discard-unreferenced-syms` above) and call the continuation function with
-;;   the /new/ slimmed expression, and a sorted-by-symbol list of binding pairs.
-;;
-;; These two return values satisfy the invariant we described above: the new
-;; expression, rehydrated using the symbol->subexpression map, should give us
-;; back the old expression.
-;;
-;; NOTE that the algorithm as implemented below uses a postwalk tree traversal!
-;; This is important, as it forces us to consider smaller subexpressions first.
-;; Consider some expression like:
+;; We implement the algorithm described in \[F90\]:
+;;
+;;   [Analytic Variations on the Common Subexpression Problem]
+;;   (https://algo.inria.fr/flajolet/Publications/FlSiSt90.pdf)
+;;   Philippe Flajolet, Paolo Sipala, Jean-Marc Steyaert (1990)
+;;
+;; From the abstract:
+;;
+;; > Any tree can be represented in a maximally compact form as
+;; > a directed acyclic graph where common subtrees are factored and
+;; > shared, being represented only once. Such a compaction can be
+;; > effected in linear time. It is used to save storage in
+;; > implementations of functional programming languages, as
+;; > well as in symbolic manipulation and computer algebra systems.
+;;
+;; which is exactly what we want.
+;;
+;; The heart of the algorithm is a function which computes a unique
+;; identifier (UID) for a subtree, such that isomorphic subtrees receive the
+;; same UID. If we regard each node of the tree as the root of such a
+;; subtree, the UID assignment will have the effect of partitioning
+;; the nodes of the tree into equivalence classes. We can then produce
+;; a DAG encoding of the original tree by using these class indices.
+;;
+;; The paper assumes the input trees are binary, but Emmy expressions
+;; often have sums and products in which the `+` and `*` functions are
+;; applied to long strings of arguments. We have observed that such long
+;; chains of operands can frustrate this kind of optimization, by hiding
+;; common (permuted) subsequences of operand within two similar
+;; subexpressions of a commutative operation.
+;;
+;; We begin by rewriting our input expression to avoid this. While this
+;; will likely generate a tree with more nodes, it exposes each primitive
+;; computation to the algorithm which enables a maximal factoring (at the
+;; cost, as we will see, of generating code in which operations are limited
+;; to two operands. This may produce less compact-looking compiled code,
+;; but should be perfectly transparent to the code generators of the host
+;; languages.)
+
+(defn- dissociate
+  "If `x` is an expression like `(f x1 x2 x3...)`, where `(f? f)` is true,
+  returns `(f ... (f x1 x2) x3 ...)`, i.e., we replace applications with
+  arity > 2 with left-associative nested applications of arity 2. Otherwise
+  `x` is returned."
+  [f? x]
+  (letfn [(go [x]
```

Comment:
Couple of ideas, not sure if they are helpful at all (your call either way!)

```clojure
(defn- dissociate [f? x]
  (letfn [(go [x]
              (cond (not (seq? x)) x
                    
                    ;; I am irrationally scared of laziness in cases like this, after being bitten in the simplifier
                    ;; implementation by a lazy `map`. Maybe it doesn't matter here, but you could `doall`...
                    (or (not (f? (first x)))
                        (< (count x) 4))
                    (doall (map go x))
                    
                    :else
                    ;; transduce would work here too and save an intermediate sequence... wdyt?
                    (let [[f x & xs] x]
                      (transduce (partial list f)
                                 (map go)
                                 x
                                 xs))))]
    (go x)))
```

Reply:
I think it is reasonable for this function to be lazy, which would be useful if this were one in a sequence of tree-transformations, and that might happen as the compiler gets more powerful. Looking at where the laziness is capped in this case, it's an `into []` enforced by its caller.

`transduce` hint gratefully accepted!

Reply:
```clojure
(transduce (map str) (partial list 'f) 'x [2 3 4])
(f (f (f (f x "2") "3") "4"))
```
Annoyingly an arity-1 call to f is used to finalize the reduction (grr). I think reduce is more transparent here

Reply:
@sritchie points out that `completing` would fix that problem, but I've stuck with reduce on the KISS principle